src/pages/Governance/importJsonProposal/index.ts

```diff
@@ -0,0 +1,134 @@
+import { VError } from 'errors';
+import { JsonProposal } from 'types';
+
+const importJsonProposal = async (proposalFile: File) => {
+  const contents = await proposalFile.text();
+  const jsonProposal = JSON.parse(contents);
+
+  const { meta, type: proposalType, signatures, targets, params } = jsonProposal as JsonProposal;
+
+  if (!meta) {
+    throw new VError({
+      type: 'proposal',
+      code: 'noMetaKey',
+    });
+  }
+
+  const { title, description, abstainDescription, againstDescription, forDescription } = meta;
+
+  if (!title) {
+    throw new VError({
+      type: 'proposal',
+      code: 'missingTitle',
+    });
+  }
+
+  if (!description) {
+    throw new VError({
+      type: 'proposal',
+      code: 'missingDescription',
+    });
+  }
+
+  if (!abstainDescription) {
+    throw new VError({
+      type: 'proposal',
+      code: 'missingAbstainDescription',
+    });
+  }
+
+  if (!againstDescription) {
+    throw new VError({
+      type: 'proposal',
+      code: 'missingAgainstDescription',
+    });
+  }
+
+  if (!forDescription) {
+    throw new VError({
+      type: 'proposal',
+      code: 'missingForDescription',
+    });
+  }
+
+  if (proposalType === undefined) {
+    throw new VError({
+      type: 'proposal',
+      code: 'noTypeKey',
+    });
+  }
+
+  if (!signatures) {
+    throw new VError({
+      type: 'proposal',
+      code: 'noSignaturesKey',
+    });
+  }
+
+  if (!targets) {
+    throw new VError({
+      type: 'proposal',
+      code: 'noTargetsKey',
+    });
+  }
+
+  if (!params) {
+    throw new VError({
+      type: 'proposal',
+      code: 'noParamsKey',
+    });
+  }
+
+  const actions = signatures.map((signature, idx) => {
+    const target = targets[idx] ? `${targets[idx]}` : undefined;
+    if (!target) {
+      throw new VError({
+        type: 'proposal',
+        code: 'missingTargetForSignature',
+        data: {
+          info: signature,
+        },
+      });
+    }
+
+    const rawCallData = params[idx] ? params[idx] : undefined;
+    if (!rawCallData) {
+      throw new VError({
+        type: 'proposal',
+        code: 'missingParamsForSignature',
+        data: {
+          info: signature,
+        },
+      });
+    }
+    let callData: string[] = [];
+    rawCallData.forEach(cd => {
+      if (Array.isArray(cd)) {
+        const arrStr = JSON.stringify(cd);
+        callData = [...callData, arrStr];
+      } else {
+        callData = [...callData, cd];
+      }
+    });
+
+    return {
+      signature,
+      target,
+      callData,
+    };
+  });
+
+  const formattedProposal = {
+    title,
+    description,
+    forDescription,
+    abstainDescription,
+    againstDescription,
+    proposalType,
+    actions,
+  };
```

Comment:
Wouldn't you be able to leverage the validation schema of the form to handle the validation of this object too?

You could shape the data you get from the uploaded JSON, like you do here, then pass this object to the validation schema. Depending on the errors it returns, you can then decide on what error message to throw/display.
The advantage I see is that you'd centralize validation and use the same set of rules for both the upload and manual flows. Does that make sense?